compiler/mir/mirtrees.nim

```diff
@@ -56,166 +56,138 @@ type
+              ## ``lent T`` local
 
-    mnkOpParam ## references a parameter of the enclosing region
-               ## XXX: see if it's possible to merge this with ``mnkParam``,
-               ##      by treating a procedure's body as a region itself
+    mnkField  ## declarative node only allowed in special contexts
 
     mnkLiteral ## literal data. Currently represented via a ``PNode``
+    # future direction: split into IntLit, FloatLit, and StrLit and store the
+    # values in a separate table (so that MirNode gets smaller)
     mnkType    ## a type literal
 
-    mnkNone    ## the "nothing" value. Represents the absence of a value.
-               ## To ease the back-to-``PNode`` translation, this node is
-               ## currently allowed to have a non-nil type.
-
-    mnkDef       ## defines an entity and, if the entity is a location, starts
-                 ## its lifetime (if the entity is a location). If the entity
-                 ## is a location, the ``def`` can be either a *sink* or a
-                 ## statement -- it must be a statement otherwise
-    mnkDefCursor ## starts the lifetime of a location that is non-owning. The
-                 ## location may or may not contain a value and is not
-                 ## responsible for destroying it
-                 ## inputs: either a single value (no arg-block) or none
-    mnkDefUnpack ## starts the lifetime of a location that owns a *tuple*
-                 ## value but is not responsible for destroying it, as all
-                 ## sub-values are moved out of it
-                 ## **Warning**: this is a hack and it should be relied on as
-                 ## little as possible. It's a workaround to support the
-                 ## temporary tuple values introduced for destructuring.
-
-    mnkFastAsgn ## ``fastAsgn(dst, src)``; assigns the `src` value to the location
-                ## named by the lvalue `dst`. Neither the previous value in the
-                ## destination location nor the source value are mutated in any
-                ## way. No transfer of ownership happens.
-    mnkAsgn     ## ``asgn(dst, src)``; assigns the `src` value to the location
-                ##  named by `dst`, also transferring onwership.
-    mnkInit     ## ``init(dst, src)``; similar to `asgn`, but with the
-                ## guarantee that the destination contains no value prior
-
-    mnkSwitch ## ``switch(x, y)``; changes the active branch of the record-case
-              ## identified by the result of the ``pathVariant`` operation used
-              ## as the `x` operand. `y` is the new discriminator value
+    # future direction:
+    # store the type of the destination within each def, assignment, etc. and
+    # then remove the type field from ``MirNode``
+
+    mnkDef       ## marks the start of existence of a local, global, procedure,
+                 ## or temporary. Supports an optional intial value (except for
+                 ## procedure definitions)
```

Comment:
I think it might be better to specialize the Def-with-value variant to flatten the conditionals.

Reply:
I would keep it as-is for now, as a separate node kind would introduce more special casing than it'd remove, at least at the moment.

Once the `mnkNone`-as-an-argument usage is removed, and `mnkDefUnpack` and `mnkDefCursor` are gone too, I'd take another look at splitting the node kind into two.